pandas/core/arrays/_utils.py

```diff
@@ -0,0 +1,54 @@
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING,
+    Any,
+)
+
+import numpy as np
+
+from pandas._libs import lib
+from pandas.errors import LossySetitemError
+
+from pandas.core.dtypes.cast import np_can_hold_element
+from pandas.core.dtypes.common import is_numeric_dtype
+
+if TYPE_CHECKING:
+    from pandas._typing import (
+        ArrayLike,
+        npt,
+    )
+
+
+def to_numpy_dtype_inference(
+    arr: ArrayLike, dtype: npt.DTypeLike | None, na_value, hasna: bool
+) -> tuple[npt.DTypeLike, Any]:
+    if dtype is None and is_numeric_dtype(arr.dtype):
+        dtype_given = False
+        if hasna:
+            if arr.dtype.kind == "b":
+                dtype = np.dtype(np.object_)
+            else:
+                if arr.dtype.kind in "iu":
+                    dtype = np.dtype(np.float64)
+                else:
+                    dtype = arr.dtype.numpy_dtype  # type: ignore[union-attr]
+                if na_value is lib.no_default:
+                    na_value = np.nan
+        else:
+            dtype = arr.dtype.numpy_dtype  # type: ignore[union-attr]
+    elif dtype is not None:
+        dtype = np.dtype(dtype)
+        dtype_given = True
+    else:
+        dtype_given = True
+
+    if na_value is lib.no_default:
+        na_value = arr.dtype.na_value
+
+    if not dtype_given and hasna:
+        try:
+            np_can_hold_element(dtype, na_value)  # type: ignore[arg-type]
+        except LossySetitemError:
+            dtype = np.dtype(np.object_)
```

Comment:
For which cases is this needed? If dtype_given=False, we know `dtype` will either be object or float-like, which should be able to hold NaN?

Reply:
bool

Reply:
Ah, I see it is copied from elsewhere. And of course `na_value` is not necessarily NaN, but can be any user provided value? But still, I wouldn't expect us to fall back to object dtype in that case, but rather raise an error (if the user provided an na_value that is not compatible with dtype they also specified themselves)

Reply:
the issue is that pd.NA is not compatible and I don't want this to raise since this is a valid request

Reply:
Sorry, no, this is of course the case where the user did _not_ provide an explicit dtype .. So it makes some sense to use object dtype if the na_value doesn't match the default numpy equivalent. 

So ignore this ;)